### ISSUE TITLE
Cronjob completed pods not deleted (job deleted)

### ISSUE THREAD

**Issue by bowang-666** [kubernetes/kubernetes#94293]
**What happened**:
Completed pods of cronjob were not deleted, remain forever. The job was deleted successfully.
Garbage collector seems stop working at certain point.

**What you expected to happen**:
Completed pods are deleted along with the parent jobs.

**How to reproduce it (as minimally and precisely as possible)**:
In a k8s cluster, run a bunch of cronjobs. (we have around 30).
This seems to happen with large work load.

**Anything else we need to know?**:
successfuljobshistorylimit is set to 1, and the job was deleted successfully:
PreciseTimeStamp | Log
-- | --
2020-08-26T01:00:35Z | I0826 01:00:35.941152 1 event.go:258] Event(v1.ObjectReference{Kind:"CronJob", Namespace:"kube-log", Name:"elasticsearch-curator", UID:"55af668c-82ca-45ff-85aa-df7c254ea6c7", APIVersion:"batch/v1beta1", ResourceVersion:"7318211", FieldPath:""}): type: 'Normal' reason: **'SuccessfulDelete' Deleted job elasticsearch-curator-1598317200**

However the corresponding pod still remain:
$ kubectl get pods -n kube-log
NAME                                     READY   STATUS      RESTARTS   AGE
elasticsearch-client-86f5c6f8d5-fgf54    1/1     Running     0          14d
elasticsearch-client-86f5c6f8d5-nxfqs    1/1     Running     0          14d
**elasticsearch-curator-1598317200-8kjm4   0/1     Completed   0          42h**
elasticsearch-curator-1598403600-2mklq   0/1     Completed   0          18h
elasticsearch-data-0                     1/1     Running     0          14d
elasticsearch-data-1                     1/1     Running     0          14d
elasticsearch-data-2                     1/1     Running     0          14d
elasticsearch-master-0                   1/1     Running     0          14d
elasticsearch-master-1                   1/1     Running     0          14d
elasticsearch-master-2                   1/1     Running     0          14d

The garbage collector seems stopped working after 2020-08-25T10:48:33Z, after that it stops garbage collecting pods:

PreciseTimeStamp | Log
-- | --
2020-08-25T10:47:13Z | I0825 10:47:13.683407       1 gc_controller.go:125] garbage collecting 2 pods
2020-08-25T10:47:13Z | E0825 10:47:13.956000       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:47:13Z | E0825 10:47:13.956038       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21105)
2020-08-25T10:47:14Z | I0825 10:47:14.455557       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
2020-08-25T10:47:33Z | I0825 10:47:33.718584       1 gc_controller.go:125] garbage collecting 2 pods
2020-08-25T10:47:44Z | E0825 10:47:44.455853       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:47:44Z | E0825 10:47:44.455889       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21106)
2020-08-25T10:47:44Z | I0825 10:47:44.955676       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
2020-08-25T10:47:53Z | I0825 10:47:53.748496       1 gc_controller.go:125] garbage collecting 7 pods
2020-08-25T10:48:14Z | E0825 10:48:14.955867       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:48:14Z | E0825 10:48:14.955905       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21107)
2020-08-25T10:48:15Z | I0825 10:48:15.455857       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
**2020-08-25T10:48:33Z** | **I0825 10:48:33.808278       1 gc_controller.go:125] garbage collecting 1 pods**
2020-08-25T10:48:45Z | E0825 10:48:45.463337       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:48:45Z | E0825 10:48:45.463377       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21108)
2020-08-25T10:48:45Z | I0825 10:48:45.956202       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
2020-08-25T10:49:15Z | E0825 10:49:15.956411       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:49:15Z | E0825 10:49:15.956455       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21109)
2020-08-25T10:49:16Z | I0825 10:49:16.455771       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
2020-08-25T10:49:46Z | E0825 10:49:46.455999       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:49:46Z | E0825 10:49:46.456029       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21110)
2020-08-25T10:49:46Z | I0825 10:49:46.955711       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
2020-08-25T10:50:16Z | E0825 10:50:16.955893       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:50:16Z | E0825 10:50:16.955958       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21111)
2020-08-25T10:50:17Z | I0825 10:50:17.455491       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller
2020-08-25T10:50:47Z | E0825 10:50:47.455753       1 controller_utils.go:1032] unable to sync caches for garbage collector controller
2020-08-25T10:50:47Z | E0825 10:50:47.455785       1 garbagecollector.go:228] timed out waiting for dependency graph builder sync during GC sync (attempt 21112)
2020-08-25T10:50:47Z | I0825 10:50:47.955418       1 controller_utils.go:1029] Waiting for caches to sync for garbage collector controller

Restarted kube-controller-manager but didn't work.

I couldn't find out what happened to GC at that time, nor what to do to bring it back. Could you advise where to look at further?

**Environment**:
- Kubernetes version (use `kubectl version`):
1.15.11
- Cloud provider or hardware configuration:
Azure Kubernetes Service
- OS (e.g: `cat /etc/os-release`):
Ubuntu 1604.6 LTS
- Kernel (e.g. `uname -a`):
4.15.0-1091-azure
- Install tools:
elasticsearch-curator (not sure if it's related)
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by azalio** [comment]
We have similar problem with v1.15.12

**Comment by mlycore** [comment]
In my opinion, that's `timed out waiting for dependency graph builder sync during GC sync (attempt 21111)`  issue. Kube-controller-manager couldn't find out the dependency of your Job and Pod. It's weird, but maybe you could check Pod.OwnerRefernece and Finalizer settings.

**Comment by azalio** [comment]
> In my opinion, that's `timed out waiting for dependency graph builder sync during GC sync (attempt 21111)` issue. Kube-controller-manager couldn't find out the dependency of your Job and Pod. It's weird, but maybe you could check Pod.OwnerRefernece and Finalizer settings.

In my case, I have checked that.
I think too it was too many resources that caused a timeout.

**Comment by ahmadbabaei** [comment]
I have the same issue but with the deployment. isn't there any solution for it yet?

**Comment by mlycore** [comment]
> I have the same issue but with the deployment. isn't there any solution for it yet?

Do you mean you cannot delete Pod after delete the Deployment ? The cascade thing?

**Comment by ahmadbabaei** [comment]
> > I have the same issue but with the deployment. isn't there any solution for it yet?
> 
> Do you mean you cannot delete Pod after delete the Deployment ? The cascade thing?

Yes exactly. The deployment deletion won't delete the pods. They can be scaled up and down.

**Comment by bcho** [comment]
We have a cluster observed similar issue as this case. Our setup flow is:

1. create a deployment
2. deployment creates the replica set, then pod
3. delete the deployment
4. deployment deleted, but the rs and pod are still remaining in the cluster

From KCM's log, we can see many log line like this:

>  timed out waiting for dependency graph builder sync during GC sync

However we didn't reproduce it in other clusters with similar flow (and we didn't see the cache sync logs like above in those clusters).

**Comment by jfpucheu** [comment]
Hello All, 

We got the similar issue on a cluster of 6000 pods.

- Deployment deleted but ReplicaSet Still Here.
- ReplicaSet deleted but Pod Still Here.

The issue come from the garbage Collector of kube-controller-manager . When too many Object on database , the garbage collector can't Get all objects on api to see what should be deleted.

In Our case , this log in controller logs give us a clue:
```
E0402 10:52:00.858591       8 shared_informer.go:226] unable to sync caches for garbage collector
E0402 10:52:00.858600       8 garbagecollector.go:233] timed out waiting for dependency graph builder sync during GC sync (attempt 16)
I0402 10:52:00.883519       8 graph_builder.go:272] garbage controller monitor not yet synced: dex.coreos.com/v1, Resource=authrequests
```
A CustomRessourceDefinition ( Crd ) was created by Dex :
kubectl get crd | grep dex
....
authrequests.dex.coreos.com 
....
and on each auth request an authrequests.dex.coreos.com  object was stored in Kubernetes Storage -> and finally in Etcd.

We discovered about 2449728 authrequests.dex.coreos.com  objects in our ETCD database.

That cause ETCD become Slow, witch cause API server Consuming 50 giga of ram and crashing ... and the GC cannot list objects to purge them ...

We delete the CRD: authrequests.dex.coreos.com , all the objects where purges, we clean and defrag our ETCD database (4 giga of data cleaned ) and all became to normal.. ( after 2 month of investigation on why our API servers was overloaded ).

Now we have to understand why our CRD are not cleaned by the Controller garbage Collector...

To conclude , Check your CRD volumes 👍  !
Jeff
